### ISSUE TITLE
`reportUnusedFunction` false negative on protected methods of `@final` classes

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#548]
Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBALgngBwJYDsDmUkQWEMrCoCGANgLABQlAJgKbBQD6wYYAFAJQBcUAxFLRDgQlKGKgA6KZUoABQilKUAxiSIBndVABirLqPF0GzVm3W0SwbpKl8oKMAKG4gA)

```python
from typing import final

def _foo(): # error
    ...

@final
class Foo:
    def _foo(self): ... # no error
```

**Comment by DetachHead** [comment]
for backwards compatibility with pyright we should probably just make a separate rule for this: `reportUnusedMethod`

**Comment by DetachHead** [comment]
upstream issue: https://github.com/microsoft/pyright/issues/4951

the reason it was rejected was because a subclass could use it, but in that case why is the method even defined on the class at all if it's only used by subclasses? for that reason i think it should be an error regardless. also it should at least work on `@final` classes
